### PR TITLE
Resource management for macro/plugin classloaders, classpath JARs

### PR DESCRIPTION
**Pull request by retronym**
Global now has a close() method, which closes any classloaders
and open JARs created for compiler plugins, macros, and the
classpath.

When these aren't owned by Global (because caching is turned on),
this reference counts are decremented on Global.close.

When the reference count for a resource reaches zero, it is queued
for close after a short (and configurable) delay.

Also introduces a setting to provide a restricted classpath for
the macro classloader, which can improve the hit ratio when
-Ycache-macro-classloader.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/GenericRunnerSettings.scala [hunk 1]
```diff
@@ -16,7 +16,14 @@ import java.net.URL
 import scala.tools.util.PathResolver
 
 class GenericRunnerSettings(error: String => Unit) extends Settings(error) {
-  lazy val classpathURLs: Seq[URL] = new PathResolver(this).resultAsURLs
+  lazy val classpathURLs: Seq[URL] = {
+    val registry = new CloseableRegistry
+    try {
+      new PathResolver(this, new CloseableRegistry).resultAsURLs
```
Comment (viktorklang) [inline src/compiler/scala/tools/nsc/GenericRunnerSettings.scala]:
```suggestion
      new PathResolver(this, registry).resultAsURLs
```

File: src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala [hunk 2]
```diff
@@ -215,3 +242,26 @@ final class FileBasedCache[T] {
     cache.clear()
   }
 }
+
+object FileBasedCache {
+  private val deferCloseMs = Integer.getInteger("scalac.filebasedcache.defer.close.ms", 1000)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala:252]: The tension here is that too long a delay could lead to an error (on Windows) with an inabilty to overwrite the JAR. To short a delay and the entry could be evicted before a subsequent sub-project compilation is able to get a cache hit. A more comprehensive solution would be to involve build tools in the policy: they could close entries with refcount of zero when that entry's JAR is about to be overwritten.

**Comment by jvican** [comment]
I'm reviewing this tomorrow, sorry for the delay!

**Review by jvican — COMMENTED** [review]
LGTM :+1:

**Comment by SethTisue** [comment]
@retronym is this still WIP?

**Comment by retronym** [comment]
@lrytz Could you please take a look at this one?

**Review by lrytz — APPROVED** [review]

**Pull request merged**